File: rust/2021/src/bin/day_08.rs
use std::fs;
use std::collections::{HashMap, HashSet};

type Signal = HashSet<u8>;

fn count_easy(entry: &str) -> usize {
    let (_, output) = entry.split_once(" | ").unwrap();
    output.split(' ').filter(|w| [2, 4, 3, 7].contains(&w.len())).count()
}

fn parse_entry(entry: &str) -> usize {
    let (input, output) = entry.split_once(" | ").unwrap();
    let mut len2sig: HashMap<usize, Vec<Signal>> = HashMap::new();
    for token in input.split(' ') {
        len2sig.entry(token.len()).or_default().push(token.bytes().collect());
    }
    // Start with the numbers of unique length
    let mut sigs: HashMap<usize, Signal> = [
        (1, len2sig[&2][0].clone()), (4, len2sig[&4][0].clone()),
        (7, len2sig[&3][0].clone()), (8, len2sig[&7][0].clone()),
    ].into_iter().collect();
    // Derive the rest
    let adg = len2sig[&5].iter().fold(sigs[&8].clone(), |acc, sig| &acc & sig);
    sigs.insert(3, &sigs[&1] | &adg);
    sigs.insert(9, &sigs[&4] | &adg);
    let be = &sigs[&8] - &sigs[&3];
    let d = &(&sigs[&4] - &sigs[&1]) - &be;
    sigs.insert(0, &sigs[&8] - &d);
    sigs.insert(6, len2sig[&6].iter().filter(|&s| *s != sigs[&0] && *s != sigs[&9]).next().unwrap().clone());
    sigs.insert(5, &sigs[&9] & &sigs[&6]);
    let ce = &sigs[&9] ^ &sigs[&6];
    sigs.insert(2, &adg | &ce);
    // Parse the output signals
    let digits: Vec<_> = output.split(' ').map(|s| {
        let signal: Signal = s.bytes().collect();
        for (num, sig) in &sigs {
            if *sig == signal {
                return *num;
            }
        }
        panic!("Unrecognized signal")
    }).collect();
    digits.iter().fold(0, |acc, &b| 10 * acc + b)
}

fn main() {
    let input = fs::read_to_string("../input/2021/input_08.txt").unwrap();
    println!("Part 1: {}", input.lines().map(count_easy).sum::<usize>());
    println!("Part 2: {}", input.lines().map(parse_entry).sum::<usize>());
}

File: haskell/2021/day_08.hs
import qualified Data.List.Split as Split
import qualified Data.Map as Map
import qualified Data.Set as Set

countEasy :: String -> Int
countEasy entry =
    let _:output:_ = Split.splitOn " | " entry
    in length [w | w <- words output, length w `elem` [2,4,3,7]]

parseEntry :: String -> Int
parseEntry entry =
    let input:output:_ = [Set.fromList <$> words s | s <- Split.splitOn " | " entry]
        setsOfLen n = [s | s <- input, Set.size s == n]
        -- Start with the signals of unique length
        one:_ = setsOfLen 2
        four:_ = setsOfLen 4
        seven:_ = setsOfLen 3
        eight:_ = setsOfLen 7
        -- Derive the rest
        adg = foldl1 Set.intersection $ setsOfLen 5
        three = Set.union one adg
        nine = Set.union four adg
        be = Set.difference eight three
        d = foldl1 Set.difference [four, one, be]
        zero = Set.difference eight d
        six:_ = [s | s <- setsOfLen 6, s `notElem` [zero, nine]]
        five = Set.intersection six nine
        ce = Set.difference eight five
        two = Set.union adg ce
        sigs = Map.fromList $ zip [zero, one, two, three, four, five, six, seven, eight, nine] [0..]
    in foldl (\acc s -> 10 * acc + (sigs Map.! s)) 0 output

main :: IO ()
main = do
    text <- readFile "../../input/2021/input_08.txt"
    putStrLn $ "Part 1: " ++ show (sum $ countEasy <$> lines text)
    putStrLn $ "Part 2: " ++ show (sum $ parseEntry <$> lines text)
